src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -255,8 +256,56 @@ private void RecordDependency(CallGraphNode callerKey, CallGraphNode calledKey,
         /// </summary>
         public List<ImmutableArray<CallGraphNode>> GetCallCycles()
         {
-            // ToDo: need to implement a robust algorithm for finding cycles
-            throw new NotImplementedException();
+            var callStack = new Dictionary<CallGraphNode, CallGraphNode>();
+            var finished = new HashSet<CallGraphNode>();
+            var cycles = new List<ImmutableArray<CallGraphNode>>();
+
+            void processDependencies(CallGraphNode node)
+            {
+                if (_Dependencies.TryGetValue(node, out var dependencies))
+                {
+                    foreach (var dependency in dependencies)
+                    {
+                        var (curr, _) = dependency;
+
```

Comment:
Style: In general, I have seen that lines are only left blank before a comment or after a block of code enclosed in {} (eg. if statements, cycles, methods). In order to have a consistent and easier to read style across the code base, please consider removing blank lines that do not fit the description that I mentioned.

Reply:
I don't think there is any rule against blank lines; it is used commonly to separate sections of code, see [here](https://github.com/microsoft/qsharp-compiler/blob/master/src/QsCompiler/CompilationManager/CompilationUnit.cs#L715) and [here](https://github.com/microsoft/qsharp-compiler/blob/master/src/QsCompiler/Core/SymbolTable.fs#L1224) for example.

Reply:
I think I wrote my comment in a way that sounded like a rule, sorry about that. Regarding blank lines, my recommendation in general is to be consistent with its use (e.g. do not add a blank space after a comment if in general blank lines are not added in that case), and try to not over use blank lines because I think it makes the code a bit less readable if there are a lot of them (however this can be subjective so maybe some people find the code more readable).

I don't think there is a need for a rule on blank lines but my main point is to be consistent with their use :)